issue_12/script.js

```diff
@@ -0,0 +1,77 @@
+function startLoading() {
+    const loadingImg = document.createElement('img');
+    loadingImg.src = 'loading-circle.gif';
+    loadingImg.alt = 'ローディング画像';
+    document.getElementById('js-loading').appendChild(loadingImg);
+}
+
+function removeLoading() {
+    document.getElementById('js-loading').remove();
+}
+
+function renderList(resolvedData) {
+    const ul = document.getElementById('js-ul');
+    const fragment = document.createDocumentFragment();
+    for (const item of resolvedData) {
+        const li = document.createElement('li');
+        const a = document.createElement('a');
+        const img = document.createElement('img');
+
+        a.href = item.a;
+        a.textContent = item.text;
+        img.src = item.img;
+        img.alt = item.alt;
+
+        fragment.appendChild(li).appendChild(a).appendChild(img);
+    }
+    ul.appendChild(fragment);
+}
+
+function renderErrorMessage(messageText) {
+    document.getElementById('js-ul').textContent = messageText;
+}
+
+async function fetchData() {
+    try {
+        const url = 'https://api.json-generator.com/templates/NluaaELSLhVe/data?access_token=ambjpfxjl00e50wa639kwndq1ofq3iuykdrv98ge';
+        const response = await fetch(url);
+        if (!response.ok) {
+            throw new Error('サーバーエラーです');
+        }
+        const result = await response.json();
+        return result.data;
+    } catch (responseError) {
+        console.error(responseError);
+        renderErrorMessage(responseError);
+        return false;
+    }
+}
+
+async function fetchListData() {
+    startLoading();
+    try {
+        const listData = await fetchData();
+        if (listData.length === 0) {
+            console.log('データがありません');
+            renderErrorMessage('データがありません');
+        }
+        return listData;
+    } catch (listDataError) {
+        console.error(listDataError);
+    } finally {
+        removeLoading();
+    }
+}
+
+async function showList() {
+    const obtainedData = await fetchListData();
+    if (obtainedData) {
+        renderList(obtainedData);
+    }
+}
```

Comment:
IMO
This way of processing is not wrong, but it seems a bit forced.
Why not make good use of exception handling? I recommend that you write it this way.
[sandbox:取得できない場合](https://codesandbox.io/s/qu-de-dekinaichang-he-29gv0e)

```js
async function fetchData() {
  try {
    const url =
      "https://api.json-generator.com/templates/NluaaELSLhVe/data?access_token=ambjpfxjl00e50wa639kwndq1ofq3iuykdrv98ge";
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`${response.status}:${response.text}`);
    }
    const result = await response.json();
    return result.data;
  } catch (e) {
　//ここで例外処理を行います。
    throw new Error(e);
  }
}

async function fetchListData() {
  startLoading();
  try {
   // listDataには上記の例外処理が代入されます。よってこの関数のcatchへ移行します。
    const listData = await fetchData();
    if (listData.length === 0) {
      console.log("データがありません");
      renderErrorMessage("データがありません");
    }
    return listData;
  } catch (e) {
//ここで上の例外処理が渡ってきてこの中の処理が実行されます。
    console.error(e.message); //エラーメッセージのみを引数にしています。
    renderErrorMessage(e.message);
  } finally {
    removeLoading();
  }
}
//以上のことからこの関数にif文は必要ありません。
async function showList() {
  const obtainedData = await fetchListData();
  renderList(obtainedData);
}
```

Reply:
Thank you for even suggesting a specific code!
Now that I understand the code you suggested, I have reflected it and one thing did not work.

If I omit the `if` in `showList()`, it works fine in the SandBox, but in Chrome's DevTools, `renderList()` also throws an error.
<img width="490" alt="スクリーンショット 2022-03-22 22 17 53" src="https://user-images.githubusercontent.com/65079611/159493654-4dd7fcc2-d501-4249-a6f2-ceba9b2e3c9d.png">

I checked the `obtainedData` in `console.log` and it was `undefined`. 
(This area is suspicious, but I don't have a clear reason yet. :sob:)

Therefore, I will push it with the if written once. :bow:
https://github.com/yoshikawa-akr/js-lessons/pull/12/commits/cb9650e57f6e952c021a8822063fb61f69d81ddb

Reply:
I did not confirm that it works in chrome. Sorry.
However, I think we can determine the true value of the data before that without using this function to determine the true value of the data.

I will share the code that Moriken sensei taught me when I did this task. I learned a function called init ().
[IMO](https://codesandbox.io/s/imo-w8ic88?file=/script.js)

Reply:
Thanks for sharing the code!
I see that by defining a variable in the parent function, it can be referenced in the inner function.
That made the code clearer as well. Thank you!
https://github.com/yoshikawa-akr/js-lessons/pull/12/commits/e7d3e1f55bd224f1dfb656d6dd48ab657b8689ba